### ISSUE TITLE
[feat] patch torch and huggingface for improved performance in existing applications

### ISSUE THREAD

**Issue by ClashLuke** [facebookresearch/xformers#78]
I'd love it if we could patch existing functions like `torch.nn.functional.softmax` with our faster xformers/triton implementation. This would allow users of pre-defined models from HuggingFace to simply call `xformers.patch()` and run twice as fast.\
Do you think that'd be possible, or would we have to rewrite all APIs first to match PyTorch's?

I'd propose writing a `patch()` function which replaces the appropriate PyTorch modules with our implementation at run-time. Alternative implementations are more than welcome, as my proposal wouldn't allow jumping into the source code of the new PyTorch function, rendering this change invisible to the user.

**Comment by blefaudeux** [comment]
There's an example here on how to patch existing models, in this case attention but could be generalized to something else https://github.com/facebookresearch/xformers/blob/main/docs/source/vision_transformers.ipynb. Else I've not checked the pytorch interface for softmax, but it's probably very reasonable and we could well have the same here, very doable
